File: examples/Example_45.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Diagnostics;
using PDFjet.NET;

/**
 *  Example_45.cs
 */
public class Example_45 {
    public Example_45() {
        PDF pdf = new PDF(new BufferedStream(
                new FileStream("Example_45.pdf", FileMode.Create)),
                Compliance.PDF_UA);

        Font f1 = new Font(pdf, "fonts/Droid/DroidSerif-Regular.ttf.stream");
        Font f2 = new Font(pdf, "fonts/Droid/DroidSerif-Italic.ttf.stream");
        Font f3 = new Font(pdf, "fonts/SourceCodePro/SourceCodePro-Regular.ttf.stream");

        f1.SetSize(14f);
        f2.SetSize(14f);
        f3.SetSize(10f);

        Page page = new Page(pdf, Letter.PORTRAIT);

        float w = 530f;
        float h = 13f;

        List<Field> fields = new List<Field>();
        fields.Add(new Field(   0f, new String[] {"Company", "Smart Widget Designs"}));
        fields.Add(new Field(   0f, new String[] {"Street Number", "120"}));
        fields.Add(new Field(  w/8, new String[] {"Street Name", "Oak"}));
        fields.Add(new Field(5*w/8, new String[] {"Street Type", "Street"}));
        fields.Add(new Field(6*w/8, new String[] {"Direction", "West"}));
        fields.Add(new Field(7*w/8, new String[] {"Suite/Floor/Apt.", "8W"})
                .SetAltDescription("Suite/Floor/Apartment")
                .SetActualText("Suite/Floor/Apartment"));
        fields.Add(new Field(   0f, new String[] {"City/Town", "Toronto"}));
        fields.Add(new Field(  w/2, new String[] {"Province", "Ontario"}));
        fields.Add(new Field(7*w/8, new String[] {"Postal Code", "M5M 2N2"}));
        fields.Add(new Field(   0f, new String[] {"Telephone Number", "(416) 331-2245"}));
        fields.Add(new Field(  w/4, new String[] {"Fax (if applicable)", "(416) 124-9879"}));
        fields.Add(new Field(  w/2, new String[] {"Email","jsmith12345@gmail.ca"}));
        fields.Add(new Field(   0f, new String[] {
                "Other Information","We don't work on weekends.", "Please send us an Email."}));

        new Form(fields)
                .SetLabelFont(f1)
                .SetLabelFontSize(7f)
                .SetValueFont(f2)
                .SetValueFontSize(9f)
                .SetLocation(50f, 50f)
                .SetRowLength(w)
                .SetRowHeight(h)
                .DrawOn(page);

        Dictionary<String, int> colors = new Dictionary<String, int>();
        colors["new"] = Color.red;
        colors["ArrayList"] = Color.blue;
        colors["List"] = Color.blue;
        colors["String"] = Color.blue;
        colors["Field"] = Color.blue;
        colors["Form"] = Color.blue;
        colors["Smart"] = Color.green;
        colors["Widget"] = Color.green;
        colors["Designs"] = Color.green;

        float x = 50;
        float y = 280;
        float dy = f3.GetBodyHeight();
        List<String> lines = Text.ReadLines("data/form-code-csharp.txt");
        foreach (String line in lines) {
            page.DrawString(f3, line, x, y, Color.gray, colors);
            y += dy;
        }

        pdf.Complete();
    }

    public static void Main(String[] args) {
        Stopwatch sw = Stopwatch.StartNew();
        long time0 = sw.ElapsedMilliseconds;
        new Example_45();
        long time1 = sw.ElapsedMilliseconds;
        sw.Stop();
        TextUtils.PrintDuration("Example_45", time0, time1);
    }

}   // End of Example_45.cs

File: examples/Example_45.java
/*
 * Example_45.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package examples;

import java.io.*;
import java.util.*;
import com.pdfjet.*;
import com.pdfjet.fonts.*;

/**
 * Example_45.java
 * Using the Form and Field classes to draw a shipment request. A field at x = 0
 * starts a new row, the other fields of the row start at their own x, and a
 * field with an empty label continues the value above it on a new line.
 */
public class Example_45 {
    public Example_45() throws Exception {

        PDF pdf = new PDF(
                new BufferedOutputStream(
                        new FileOutputStream("Example_45.pdf")));

        Font f1 = new Font(pdf, IBMPlexSans.Regular);
        Font f2 = new Font(pdf, IBMPlexSans.SemiBold);

        Page page = new Page(pdf, Letter.PORTRAIT);

        TextLine text = new TextLine(f2, "Shipment Request");
        text.setFontSize(22f);
        text.setLocation(56f, 80f);
        text.drawOn(page);

        TextBlock textBlock = new TextBlock(f1,
                "A Form draws its fields in rows. A field at x = 0 starts a new row, "
                + "the other fields of the row start at their own x, and a field with "
                + "an empty label continues the value above it on a new line.");
        textBlock.setFontSize(12f);
        textBlock.setLineSpacing(1.5f);
        textBlock.setLocation(56f, 95f);
        textBlock.setWidth(500f);
        float[] xy = textBlock.drawOn(page);

        float w = 500f; // The width of the form

        List<Field> fields = new ArrayList<Field>();
        fields.add(new Field(   0f, "Sender", "Maple Leaf Instruments Ltd."));
        fields.add(new Field(   0f, "Street Address", "480 King Street West"));
        fields.add(new Field(6*w/8, "Suite", "1200"));
        fields.add(new Field(   0f, "City", "Toronto"));
        fields.add(new Field(3*w/8, "Province", "Ontario"));
        fields.add(new Field(5*w/8, "Postal Code", "M5V 1L7"));
        fields.add(new Field(6*w/8, "Country", "Canada"));
        fields.add(new Field(   0f, "Recipient", "Nordic Sensor Labs AB"));
        fields.add(new Field(   0f, "Street Address", "Drottninggatan 55"));
        fields.add(new Field(6*w/8, "Floor", "3"));
        fields.add(new Field(   0f, "City", "Stockholm"));
        fields.add(new Field(5*w/8, "Postal Code", "111 21"));
        fields.add(new Field(6*w/8, "Country", "Sweden"));
        fields.add(new Field(   0f, "Contact", "Anna Lindqvist"));
        fields.add(new Field(3*w/8, "Email", "anna.lindqvist@example.com"));
        fields.add(new Field(   0f, "Contents", "Two calibrated pressure sensors"));
        fields.add(new Field(5*w/8, "Weight", "3.2 kg"));
        fields.add(new Field(6*w/8, "Declared Value", "CAD 1,450.00"));
        fields.add(new Field(   0f, "Instructions",
                "Keep upright and away from magnets. Deliver on a weekday"));
        fields.add(new Field(   0f, "", "between 9:00 and 17:00, to the reception on the third floor."));

        xy = new Form(fields)
                .setLabelFont(f1)
                .setLabelFontSize(8f)
                .setLabelColor(Color.gray)
                .setValueFont(f2)
                .setValueFontSize(10f)
                .setValueColor(Color.black)
                .setLocation(56f, xy[1] + 20f)
                .setWidth(w)
                .setStrokeWidth(0.5f)
                .drawOn(page);

        text = new TextLine(f1, "The recipient signs for the package on delivery.");
        text.setFontSize(10f);
        text.setTextColor(Color.gray);
        text.setLocation(56f, xy[1] + 20f);
        text.drawOn(page);

        pdf.complete();
    }

    public static void main(String[] args) throws Exception {
        long time0 = System.currentTimeMillis();
        new Example_45();
        long time1 = System.currentTimeMillis();
        System.out.printf("Example_45 => %4d ms%n", time1 - time0);
    }
}   // End of Example_45.java
